libudpard_demo/src/udp.c

```diff
@@ -0,0 +1,251 @@
+            self->fd = -1;
+        }
+    }
+    return res;
+}
+
+int16_t udpTxSend(UDPTxHandle* const self,
+                  const uint32_t     remote_address,
+                  const uint16_t     remote_port,
+                  const uint8_t      dscp,
+                  const size_t       payload_size,
+                  const void* const  payload)
+{
+    int16_t res = -EINVAL;
+    if ((self != NULL) && (self->fd >= 0) && (remote_address > 0) && (remote_port > 0) && (payload != NULL) &&
+        (dscp <= DSCP_MAX))
+    {
+        const int dscp_int = dscp << 2U;  // The 2 least significant bits are used for the ECN field.
+        (void) setsockopt(self->fd, IPPROTO_IP, IP_TOS, &dscp_int, sizeof(dscp_int));  // Best effort.
+        const ssize_t send_result =
+            sendto(self->fd,
+                   payload,
+                   payload_size,
+                   MSG_DONTWAIT,
+                   (struct sockaddr*) &(struct sockaddr_in){.sin_family = AF_INET,
+                                                            .sin_addr   = {.s_addr = htonl(remote_address)},
+                                                            .sin_port   = htons(remote_port)},
+                   sizeof(struct sockaddr_in));
+        if (send_result == (ssize_t) payload_size)
+        {
+            res = 1;
+        }
+        else if ((errno == EAGAIN) || (errno == EWOULDBLOCK))
+        {
+            res = 0;
+        }
+        else
+        {
+            res = (int16_t) -errno;
+        }
+    }
+    return res;
+}
+
+void udpTxClose(UDPTxHandle* const self)
+{
+    if ((self != NULL) && (self->fd >= 0))
+    {
+        (void) close(self->fd);
+        self->fd = -1;
+    }
+}
+
+int16_t udpRxInit(UDPRxHandle* const self,
+                  const uint32_t     local_iface_address,
+                  const uint32_t     multicast_group,
+                  const uint16_t     remote_port)
+{
+    int16_t res = -EINVAL;
+    if ((self != NULL) && (local_iface_address > 0) && isMulticast(multicast_group) && (remote_port > 0))
+    {
+        const int reuse = 1;
+        self->fd        = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);
+        bool ok         = self->fd >= 0;
+        // Set non-blocking mode.
+        ok = ok && (fcntl(self->fd, F_SETFL, O_NONBLOCK) == 0);
+        // Allow other applications to use the same Cyphal port as well. This must be done before binding.
+        // Failure to do so will make it impossible to run more than one Cyphal/UDP node on the same host.
+        ok = ok && (setsockopt(self->fd, SOL_SOCKET, SO_REUSEADDR, &reuse, sizeof(reuse)) == 0);
+#ifdef SO_REUSEPORT  // Linux
+        ok = ok && (setsockopt(self->fd, SOL_SOCKET, SO_REUSEPORT, &reuse, sizeof(reuse)) == 0);
+#endif
+        // Binding to the multicast group address is necessary on GNU/Linux: https://habr.com/ru/post/141021/
+        // Binding to a multicast address is not allowed on Windows, and it is not necessary there;
```

Comment:
So...this will fail on Windows?

Reply:
Yes, this will fail on Windows. To make it work there, the binding has to be conditional. I could throw in some ifdefs in there but we don't really want to test it against Windows, do we?

Reply:
It'd be nice if this worked on Windows, yes.

Reply:
I added the required preprocessor check but this is not yet tested on Windows. I'm afraid that ensuring compatibility with Windows is a rather significant chunk of work.